################################################################################
SHORT DESCRIPTION: 
################################################################################
Get bootloader MBI video information.

################################################################################
LONG DESCRIPTION: 
################################################################################
Get the MBI video information from the bootloader to Xen which will then pass it
to dom0. This is needed so early boot graphics work correctly.

################################################################################
CHANGELOG 
################################################################################

################################################################################
REMOVAL 
################################################################################

################################################################################
UPSTREAM PLAN
################################################################################

################################################################################
INTERNAL DEPENDENCIES 
################################################################################

################################################################################
PATCHES 
################################################################################
--- a/xen/arch/x86/boot/reloc.c
+++ b/xen/arch/x86/boot/reloc.c
@@ -141,7 +141,8 @@ static multiboot_info_t *mbi_reloc(u32 m
                        MBI_CMDLINE |
                        MBI_MODULES |
                        MBI_MEMMAP |
-                       MBI_LOADERNAME);
+                       MBI_LOADERNAME |
+                       MBI_FRAMEBUFFER_INFO);
 
     return mbi_out;
 }
--- a/xen/arch/x86/setup.c
+++ b/xen/arch/x86/setup.c
@@ -95,6 +95,9 @@ unsigned long __initdata highmem_start;
 size_param("highmem-start", highmem_start);
 #endif
 
+static bool_t __initdata mbi_video;
+boolean_param("mbi-video", mbi_video);
+
 cpumask_t __read_mostly cpu_present_map;
 
 unsigned long __read_mostly xen_phys_start;
@@ -507,7 +510,7 @@ struct boot_video_info {
 extern struct boot_video_info boot_vid_info;
 #endif
 
-static void __init parse_video_info(void)
+static void __init parse_video_info(multiboot_info_t *mbi)
 {
 #ifdef CONFIG_VIDEO
     struct boot_video_info *bvi = &bootsym(boot_vid_info);
@@ -516,6 +519,30 @@ static void __init parse_video_info(void
     if ( efi_enabled(EFI_BOOT) )
         return;
 
+    if ( mbi_video &&
+        (mbi->flags & MBI_FRAMEBUFFER_INFO) &&
+        (mbi->framebuffer_type==MULTIBOOT_FRAMEBUFFER_TYPE_RGB)) {
+       bvi->orig_video_isVGA = 0x23;
+
+       bvi->lfb_width          = mbi->framebuffer_width;
+       bvi->lfb_height         = mbi->framebuffer_height;
+       bvi->lfb_linelength     = mbi->framebuffer_pitch;
+       bvi->lfb_depth          = mbi->framebuffer_bpp;
+       bvi->lfb_base           = mbi->framebuffer_addr;
+       bvi->lfb_size           = mbi->framebuffer_height*mbi->framebuffer_pitch;
+
+       bvi->red_pos            = mbi->framebuffer_red_field_position;
+       bvi->red_size           = mbi->framebuffer_red_mask_size;
+       bvi->green_pos          = mbi->framebuffer_green_field_position;
+       bvi->green_size         = mbi->framebuffer_green_mask_size;
+       bvi->blue_pos           = mbi->framebuffer_blue_field_position;
+       bvi->blue_size          = mbi->framebuffer_blue_mask_size;
+       bvi->rsvd_pos           = 0;
+       bvi->rsvd_size          = 0;
+       bvi->capabilities       = 0;
+       bvi->vesa_attrib        = 0;
+    }
+
     if ( (bvi->orig_video_isVGA == 1) && (bvi->orig_video_mode == 3) )
     {
         vga_console_info.video_type = XEN_VGATYPE_TEXT_MODE_3;
@@ -760,7 +787,7 @@ void __init noreturn __start_xen(unsigne
 
     probe_hypervisor();
 
-    parse_video_info();
+    parse_video_info(mbi);
 
     rdmsrl(MSR_EFER, this_cpu(efer));
     asm volatile ( "mov %%cr4,%0" : "=r" (get_cpu_info()->cr4) );
--- a/xen/include/xen/multiboot.h
+++ b/xen/include/xen/multiboot.h
@@ -25,7 +25,6 @@
 #define MULTIBOOT_HEADER_MAGIC         0x1BADB002
 #define MULTIBOOT_HEADER_MODS_ALIGNED  0x00000001
 #define MULTIBOOT_HEADER_WANT_MEMORY   0x00000002
-#define MULTIBOOT_HEADER_HAS_VBE       0x00000004
 #define MULTIBOOT_HEADER_HAS_ADDR      0x00010000
 
 /* The magic number passed by a Multiboot-compliant boot loader. */
@@ -42,7 +41,7 @@
 #define MBI_BIOSCONFIG (_AC(1,u) << 8)
 #define MBI_LOADERNAME (_AC(1,u) << 9)
 #define MBI_APM        (_AC(1,u) << 10)
-
+#define MBI_FRAMEBUFFER_INFO (_AC(1,u) <<12)
 #ifndef __ASSEMBLY__
 
 /* The symbol table for a.out.  */
@@ -101,6 +100,41 @@ typedef struct {
 
     /* Valid if flags sets MBI_APM */
     u32 apm_table;
+
+    /* Video */
+    u32 vbe_control_info;
+    u32 vbe_mode_info;
+    u16 vbe_mode;
+    u16 vbe_interface_seg;
+    u16 vbe_interface_off;
+    u16 vbe_interface_len;
+
+    u64 framebuffer_addr;
+    u32 framebuffer_pitch;
+    u32 framebuffer_width;
+    u32 framebuffer_height;
+    u8 framebuffer_bpp;
+#define MULTIBOOT_FRAMEBUFFER_TYPE_INDEXED 0
+#define MULTIBOOT_FRAMEBUFFER_TYPE_RGB     1
+#define MULTIBOOT_FRAMEBUFFER_TYPE_EGA_TEXT     2
+    u8 framebuffer_type;
+    union
+    {
+      struct
+      {
+        u32 framebuffer_palette_addr;
+        u16 framebuffer_palette_num_colors;
+      };
+      struct
+      {
+        u8 framebuffer_red_field_position;
+        u8 framebuffer_red_mask_size;
+        u8 framebuffer_green_field_position;
+        u8 framebuffer_green_mask_size;
+        u8 framebuffer_blue_field_position;
+        u8 framebuffer_blue_mask_size;
+      };
+    };
 } multiboot_info_t;
 
 /* The module structure.  */
